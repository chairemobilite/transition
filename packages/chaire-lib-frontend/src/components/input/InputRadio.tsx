/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';

interface choiceType {
    value: string | boolean;
    [key: string]: unknown;
}

export interface InputRadioProps {
    id: string;
    onValueChange?: (e: any) => void;
    value?: string | boolean;
    defaultValue?: string;
    choices: choiceType[];
    sameLine?: boolean;
    isBoolean?: boolean;
    localePrefix?: string;
    t?: (string) => string;
    disabled?: boolean;
}

interface defaultInputType {
    name: string;
    disabled?: boolean;
}

// TODO: Make a real boolean component

class InputRadio extends React.Component<InputRadioProps> {
    static defaultProps: Partial<InputRadioProps> = {
        onValueChange: undefined,
        value: undefined,
        defaultValue: '',
        choices: [],
        localePrefix: 'main',
        t: (_str) => '',
        sameLine: true,
        isBoolean: false,
        disabled: false
    };

    constructor(props: InputRadioProps) {
        super(props);

        this.handleRadioLabelClick = this.handleRadioLabelClick.bind(this);
        this.handleRadioClick = this.handleRadioClick.bind(this);
        this.handleRadioContainerClick = this.handleRadioContainerClick.bind(this);
        this.handleRadioChange = this.handleRadioChange.bind(this);
    }

    handleRadioLabelClick(e: React.MouseEvent): void {
        e.stopPropagation();
    }

    handleRadioClick(e: React.MouseEvent): void {
        e.stopPropagation();
    }

    handleRadioContainerClick(inputRadioRef: React.RefObject<HTMLInputElement>, e: React.MouseEvent): void {
        e.stopPropagation();
        const clickEvent = new MouseEvent('click', {
            view: window,
            bubbles: true,
            cancelable: false
        });
        if (inputRadioRef.current) {
            inputRadioRef.current.dispatchEvent(clickEvent);
        }
    }

    handleRadioChange(e: any): void {
        e.stopPropagation();
        const onChange = this.props.onValueChange;
        if (onChange) {
            let eventValue: any = e;
            if (this.props.isBoolean) {
                eventValue = { target: { value: e.target.value.toString() === 'true' } };
            }
            onChange(eventValue);
        }
    }

    render(): React.ReactNode {
        const defaultAttributes: defaultInputType = {
            name: this.props.id
        };
        if (this.props.disabled === true) {
            defaultAttributes.disabled = true;
        }

        let value = this.props.value === undefined ? this.props.defaultValue : this.props.value;
        if (!value) {
            value = this.props.isBoolean ? 'false' : '';
        }
        if (this.props.isBoolean) {
            value = value.toString();
        }

        const radioChoices: JSX.Element[] = this.props.choices.map((choice) => {
            const inputRadioRef: React.RefObject<HTMLInputElement> = React.createRef();
            const valueStr = choice.value.toString();
            const id = `${this.props.id}_${valueStr}`;
            return (
                <div
                    className={`tr__form-input-radio-container${valueStr === value ? ' checked' : ''}`}
                    key={valueStr}
                    onClick={(e) => {
                        this.handleRadioContainerClick(inputRadioRef, e);
                    }}
                >
                    <input
                        {...defaultAttributes}
                        autoComplete="none"
                        type="radio"
                        className="tr__form-input-radio apptr__input _input-radio"
                        id={id}
                        checked={valueStr === value}
                        onChange={this.handleRadioChange}
                        onClick={this.handleRadioClick}
                        value={valueStr}
                    />
                    <label htmlFor={id} onClick={this.handleRadioLabelClick}>
                        <span>
                            {(this.props.t && this.props.t(`${this.props.localePrefix || 'main'}:${valueStr}`)) ||
                                valueStr}
                        </span>
                    </label>
                </div>
            );
        });

        return (
            <div
                className={`tr__form-input-radio-group-container _input-proxy${
                    this.props.sameLine === false ? ' no-wrap' : ''
                }`}
            >
                {radioChoices}
            </div>
        );
    }
}

export default InputRadio;
