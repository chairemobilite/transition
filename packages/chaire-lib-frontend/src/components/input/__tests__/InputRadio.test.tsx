/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import InputRadio from '../InputRadio';

const mockOnChange = jest.fn();
const testId = 'RadioWidgetId';
const defaultChoiceValue = 'test1';
const anotherChoiceValue = 'test2';
const testChoices = [
    { value: defaultChoiceValue },
    { value: anotherChoiceValue },
    { value: 'test3', unused: 'unused' },
];
const booleanChoices = [
    { value: 'true' },
    { value: 'false' }
];

test('Default props', () => {
    const { container } = render(<InputRadio
        id = {testId}
    />);
    expect(container).toMatchSnapshot();
});

test('All props', () => {
    const { container } = render(<InputRadio
        id = {testId}
        onValueChange = {mockOnChange}
        value = {anotherChoiceValue}
        defaultValue = {defaultChoiceValue}
        choices = {testChoices}
        localePrefix = "something"
        t = {(str) => str}
        sameLine = {true}
        isBoolean = {false}
        disabled = {false}
    />);
    expect(container).toMatchSnapshot();
});

test('Disabled', () => {
    const { container } = render(<InputRadio
        id = {testId}
        onValueChange = {mockOnChange}
        choices = {testChoices}
        value = {anotherChoiceValue}
        disabled = {true}
    />);
    expect(container).toMatchSnapshot();
});

test('On same line', () => {
    const { container } = render(<InputRadio
        id = {testId}
        onValueChange = {mockOnChange}
        choices = {testChoices}
        sameLine = {true}
    />);
    expect(container).toMatchSnapshot();
    const inputMultipleLine = render(<InputRadio
        id = {testId}
        onValueChange = {mockOnChange}
        choices = {testChoices}
        sameLine = {false}
    />);
    expect(inputMultipleLine).toMatchSnapshot();
});

test('Boolean choices', () => {
    const { container } = render(<InputRadio
        id = {testId}
        onValueChange = {mockOnChange}
        choices = {booleanChoices}
        isBoolean = {true}
    />);
    expect(container).toMatchSnapshot();
});

test('Default value', () => {
    mockOnChange.mockClear();
    const { getByLabelText } = render(
        <div>
            <InputRadio
                id = {testId}
                onValueChange = {mockOnChange}
                defaultValue = {defaultChoiceValue}
                choices = {testChoices}
            />
        </div>);
    const input = getByLabelText(defaultChoiceValue) as HTMLInputElement;
    expect(input.checked).toBe(true);
    const unselected = getByLabelText(anotherChoiceValue) as HTMLInputElement;
    expect(unselected.checked).toBe(false);
});

test('Default and initial value', () => {
    mockOnChange.mockClear();
    const { getByLabelText } = render(
        <div>
            <InputRadio
                id = {testId}
                onValueChange = {mockOnChange}
                defaultValue = {defaultChoiceValue}
                value = {anotherChoiceValue}
                choices = {testChoices}
            />
        </div>);
    const input = getByLabelText(anotherChoiceValue) as HTMLInputElement;
    expect(input.checked).toBe(true);
    const unselected = getByLabelText(defaultChoiceValue) as HTMLInputElement;
    expect(unselected.checked).toBe(false);
});
test('Invalid default and value', () => {
    mockOnChange.mockClear();
    const { getByLabelText } = render(
        <div>
            <InputRadio
                id = {testId}
                onValueChange = {mockOnChange}
                defaultValue = "not a value"
                value = "still not a choice"
                choices = {testChoices}
            />
        </div>);
    let input = getByLabelText(defaultChoiceValue) as HTMLInputElement;
    expect(input.checked).toBe(false);
    input = getByLabelText(anotherChoiceValue) as HTMLInputElement;
    expect(input.checked).toBe(false);
});

// Somehow, this test doesn't work, because it's a radio and they're fancy?
test.skip('Call onChange', () => {
    mockOnChange.mockClear();
    const { getByLabelText } = render(
        <div>
            <InputRadio
                id = {testId}
                onValueChange = {mockOnChange}
                choices = {testChoices}
            />
        </div>);
    const input = getByLabelText(anotherChoiceValue) as HTMLInputElement;
    expect(input.checked).toBe(false);
    fireEvent.change(input, { target: { value: anotherChoiceValue } });
    expect(mockOnChange).toHaveBeenCalledTimes(1);
});
