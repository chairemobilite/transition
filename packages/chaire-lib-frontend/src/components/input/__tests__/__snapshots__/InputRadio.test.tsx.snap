// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All props 1`] = `
<div>
  <div
    class="tr__form-input-radio-group-container _input-proxy"
  >
    <div
      class="tr__form-input-radio-container"
    >
      <input
        autocomplete="none"
        class="tr__form-input-radio apptr__input _input-radio"
        id="RadioWidgetId_test1"
        name="RadioWidgetId"
        type="radio"
        value="test1"
      />
      <label
        for="RadioWidgetId_test1"
      >
        <span>
          something:test1
        </span>
      </label>
    </div>
    <div
      class="tr__form-input-radio-container checked"
    >
      <input
        autocomplete="none"
        checked=""
        class="tr__form-input-radio apptr__input _input-radio"
        id="RadioWidgetId_test2"
        name="RadioWidgetId"
        type="radio"
        value="test2"
      />
      <label
        for="RadioWidgetId_test2"
      >
        <span>
          something:test2
        </span>
      </label>
    </div>
    <div
      class="tr__form-input-radio-container"
    >
      <input
        autocomplete="none"
        class="tr__form-input-radio apptr__input _input-radio"
        id="RadioWidgetId_test3"
        name="RadioWidgetId"
        type="radio"
        value="test3"
      />
      <label
        for="RadioWidgetId_test3"
      >
        <span>
          something:test3
        </span>
      </label>
    </div>
  </div>
</div>
`;

exports[`Boolean choices 1`] = `
<div>
  <div
    class="tr__form-input-radio-group-container _input-proxy"
  >
    <div
      class="tr__form-input-radio-container"
    >
      <input
        autocomplete="none"
        class="tr__form-input-radio apptr__input _input-radio"
        id="RadioWidgetId_true"
        name="RadioWidgetId"
        type="radio"
        value="true"
      />
      <label
        for="RadioWidgetId_true"
      >
        <span>
          true
        </span>
      </label>
    </div>
    <div
      class="tr__form-input-radio-container checked"
    >
      <input
        autocomplete="none"
        checked=""
        class="tr__form-input-radio apptr__input _input-radio"
        id="RadioWidgetId_false"
        name="RadioWidgetId"
        type="radio"
        value="false"
      />
      <label
        for="RadioWidgetId_false"
      >
        <span>
          false
        </span>
      </label>
    </div>
  </div>
</div>
`;

exports[`Default props 1`] = `
<div>
  <div
    class="tr__form-input-radio-group-container _input-proxy"
  />
</div>
`;

exports[`Disabled 1`] = `
<div>
  <div
    class="tr__form-input-radio-group-container _input-proxy"
  >
    <div
      class="tr__form-input-radio-container"
    >
      <input
        autocomplete="none"
        class="tr__form-input-radio apptr__input _input-radio"
        disabled=""
        id="RadioWidgetId_test1"
        name="RadioWidgetId"
        type="radio"
        value="test1"
      />
      <label
        for="RadioWidgetId_test1"
      >
        <span>
          test1
        </span>
      </label>
    </div>
    <div
      class="tr__form-input-radio-container checked"
    >
      <input
        autocomplete="none"
        checked=""
        class="tr__form-input-radio apptr__input _input-radio"
        disabled=""
        id="RadioWidgetId_test2"
        name="RadioWidgetId"
        type="radio"
        value="test2"
      />
      <label
        for="RadioWidgetId_test2"
      >
        <span>
          test2
        </span>
      </label>
    </div>
    <div
      class="tr__form-input-radio-container"
    >
      <input
        autocomplete="none"
        class="tr__form-input-radio apptr__input _input-radio"
        disabled=""
        id="RadioWidgetId_test3"
        name="RadioWidgetId"
        type="radio"
        value="test3"
      />
      <label
        for="RadioWidgetId_test3"
      >
        <span>
          test3
        </span>
      </label>
    </div>
  </div>
</div>
`;

exports[`On same line 1`] = `
<div>
  <div
    class="tr__form-input-radio-group-container _input-proxy"
  >
    <div
      class="tr__form-input-radio-container"
    >
      <input
        autocomplete="none"
        class="tr__form-input-radio apptr__input _input-radio"
        id="RadioWidgetId_test1"
        name="RadioWidgetId"
        type="radio"
        value="test1"
      />
      <label
        for="RadioWidgetId_test1"
      >
        <span>
          test1
        </span>
      </label>
    </div>
    <div
      class="tr__form-input-radio-container"
    >
      <input
        autocomplete="none"
        class="tr__form-input-radio apptr__input _input-radio"
        id="RadioWidgetId_test2"
        name="RadioWidgetId"
        type="radio"
        value="test2"
      />
      <label
        for="RadioWidgetId_test2"
      >
        <span>
          test2
        </span>
      </label>
    </div>
    <div
      class="tr__form-input-radio-container"
    >
      <input
        autocomplete="none"
        class="tr__form-input-radio apptr__input _input-radio"
        id="RadioWidgetId_test3"
        name="RadioWidgetId"
        type="radio"
        value="test3"
      />
      <label
        for="RadioWidgetId_test3"
      >
        <span>
          test3
        </span>
      </label>
    </div>
  </div>
</div>
`;

exports[`On same line 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="tr__form-input-radio-group-container _input-proxy"
      >
        <div
          class="tr__form-input-radio-container"
        >
          <input
            autocomplete="none"
            class="tr__form-input-radio apptr__input _input-radio"
            id="RadioWidgetId_test1"
            name="RadioWidgetId"
            type="radio"
            value="test1"
          />
          <label
            for="RadioWidgetId_test1"
          >
            <span>
              test1
            </span>
          </label>
        </div>
        <div
          class="tr__form-input-radio-container"
        >
          <input
            autocomplete="none"
            class="tr__form-input-radio apptr__input _input-radio"
            id="RadioWidgetId_test2"
            name="RadioWidgetId"
            type="radio"
            value="test2"
          />
          <label
            for="RadioWidgetId_test2"
          >
            <span>
              test2
            </span>
          </label>
        </div>
        <div
          class="tr__form-input-radio-container"
        >
          <input
            autocomplete="none"
            class="tr__form-input-radio apptr__input _input-radio"
            id="RadioWidgetId_test3"
            name="RadioWidgetId"
            type="radio"
            value="test3"
          />
          <label
            for="RadioWidgetId_test3"
          >
            <span>
              test3
            </span>
          </label>
        </div>
      </div>
    </div>
    <div>
      <div
        class="tr__form-input-radio-group-container _input-proxy no-wrap"
      >
        <div
          class="tr__form-input-radio-container"
        >
          <input
            autocomplete="none"
            class="tr__form-input-radio apptr__input _input-radio"
            id="RadioWidgetId_test1"
            name="RadioWidgetId"
            type="radio"
            value="test1"
          />
          <label
            for="RadioWidgetId_test1"
          >
            <span>
              test1
            </span>
          </label>
        </div>
        <div
          class="tr__form-input-radio-container"
        >
          <input
            autocomplete="none"
            class="tr__form-input-radio apptr__input _input-radio"
            id="RadioWidgetId_test2"
            name="RadioWidgetId"
            type="radio"
            value="test2"
          />
          <label
            for="RadioWidgetId_test2"
          >
            <span>
              test2
            </span>
          </label>
        </div>
        <div
          class="tr__form-input-radio-container"
        >
          <input
            autocomplete="none"
            class="tr__form-input-radio apptr__input _input-radio"
            id="RadioWidgetId_test3"
            name="RadioWidgetId"
            type="radio"
            value="test3"
          />
          <label
            for="RadioWidgetId_test3"
          >
            <span>
              test3
            </span>
          </label>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="tr__form-input-radio-group-container _input-proxy no-wrap"
    >
      <div
        class="tr__form-input-radio-container"
      >
        <input
          autocomplete="none"
          class="tr__form-input-radio apptr__input _input-radio"
          id="RadioWidgetId_test1"
          name="RadioWidgetId"
          type="radio"
          value="test1"
        />
        <label
          for="RadioWidgetId_test1"
        >
          <span>
            test1
          </span>
        </label>
      </div>
      <div
        class="tr__form-input-radio-container"
      >
        <input
          autocomplete="none"
          class="tr__form-input-radio apptr__input _input-radio"
          id="RadioWidgetId_test2"
          name="RadioWidgetId"
          type="radio"
          value="test2"
        />
        <label
          for="RadioWidgetId_test2"
        >
          <span>
            test2
          </span>
        </label>
      </div>
      <div
        class="tr__form-input-radio-container"
      >
        <input
          autocomplete="none"
          class="tr__form-input-radio apptr__input _input-radio"
          id="RadioWidgetId_test3"
          name="RadioWidgetId"
          type="radio"
          value="test3"
        />
        <label
          for="RadioWidgetId_test3"
        >
          <span>
            test3
          </span>
        </label>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
